openssl/net_io_ssl.c

```diff
@@ -785,7 +795,12 @@ int ci_tls_connect_nonblock(ci_connection_t *connection, const char *servername,
             X509_VERIFY_PARAM *param = SSL_get0_param(ssl);
             /* Enable automatic hostname checks */
             X509_VERIFY_PARAM_set_hostflags(param, X509_CHECK_FLAG_NO_PARTIAL_WILDCARDS);
-            X509_VERIFY_PARAM_set1_host(param, servername, 0);
+            if (servername_is_ip) {
+                X509_VERIFY_PARAM_set1_ip_asc(param, servername);
```

Comment:
The only issue I am finding is that the X509_VERIFY_PARAM_set1_ip_asc is added with the openssl-1.1.0 releases, but still maybe there are instalations which uses openssl-1.0.x releases or earlier.
If you are familiar with autoconf tools, you may want to add a check for X509_VERIFY_PARAM_set1_ip_asc in configure.ac.

If you do not want to do it I wll fix it myself after merging this PR to master branch.

Reply:
``X509_VERIFY_PARAM_set1_host`` and ``X509_VERIFY_PARAM_set1_ip_asc`` have both been added in OpenSSL 1.0.2.

The call to ``X509_VERIFY_PARAM_set1_host`` was already present in the old code, so I thought that it's OK to also use ``X509_VERIFY_PARAM_set1_ip_asc``.

If you want to support OpenSSL 1.0.1 and earlier, it would be great if you could add the autoconf check.

Reply:
True. Moreover the related c-icap code is wrapped inside the `#if defined(OPENSSL_VERSION_NUMBER) && OPENSSL_VERSION_NUMBER >= 0x1000201fL` which in practice checks for openSSL releases greater than 1.0.2.
So this part of code is correct and compiles for older OpenSSL releases.

Latter in the same function, there is the following code:
```
#if !defined(OPENSSL_VERSION_NUMBER) || OPENSSL_VERSION_NUMBER <= 0x1000201fL
    if (SSL_get_verify_mode(ssl) & SSL_VERIFY_PEER) {
        X509 *cert = SSL_get_peer_certificate(ssl);
        if (!match_X509_names(cert, servername)) {
```
which checks the given servername against server certificate names for older OpenSSL releases. Adding a check for IP addresses here to support older OpenSSL releases does not looks realistic.
We can ignore it.